aiida/engine/processes/calcjobs/calcjob.py

```diff
@@ -689,17 +689,22 @@ def presubmit(self, folder: Folder) -> CalcInfo:
                 raise PluginInternalError('CalcInfo should have the information of the code to be launched')
             this_code = load_node(code_info.code_uuid, sub_classes=(Code,))
 
-            # To determine whether this code should be run with MPI enabled, we get the value that was set in the inputs
-            # of the entire process, which can then be overwritten by the value from the `CodeInfo`. This allows plugins
+            # To determine whether this code should be run with MPI enabled, we first check the value defined as
+            # the code attribute. This value has the lowest priority and can be overridden further.
+            # Next, we check the value that was set in the inputs of the entire process. If it is present, it overrides
+            # the variable. Finally, the variable can be modified by the value from the `CodeInfo`. This allows plugins
             # to force certain codes to run without MPI, even if the user wants to run all codes with MPI whenever
             # possible. This use case is typically useful for `CalcJob`s that consist of multiple codes where one or
             # multiple codes always have to be executed without MPI.
 
-            this_withmpi = self.node.get_option('withmpi')
+            this_withmpi = this_code.get_withmpi()  # Low priority value.
+
+            input_withmpi = self.node.get_option('withmpi')  # Medium priority value.
+            this_withmpi = this_withmpi if input_withmpi is None else input_withmpi
 
             # Override the value of `withmpi` with that of the `CodeInfo` if and only if it is set
             if code_info.withmpi is not None:
-                this_withmpi = code_info.withmpi
+                this_withmpi = code_info.withmpi  # High priority value.
```

Comment:
I agree with @sphuber's suggestion. If @unkcpz is not against it - I am happy to modify the logic according to it.

Reply:
Fair enough, I have no objection to this. 

>  If this really is the case, then instead of ignoring the other settings, I think the plugin should simply add a check in the prepare_for_submission and raise an error.

Only for this, is this mean the code for checking needs to be in the plugin, and the plugin developer need to do it explicitly?

Reply:
> >  If this really is the case, then instead of ignoring the other settings, I think the plugin should simply add a check in the prepare_for_submission and raise an error.

> Only for this, is this mean the code for checking needs to be in the plugin, and the plugin developer need to do it explicitly?

I think this is what @sphuber meant. And it makes sense to me. The case is so special, so it probably needs special treatment.

Reply:
Yes that is indeed what I meant. If a plugin should really never ever be run in parallel, then the plugin developer should simply put in a check in `prepare_for_submission` and raise if either `Code.withmpi` or `metadata.options.withmpi` is `True`.

Reply:
Ok, since we agreed, I am going to proceed to implement the discussed changes.

Reply:
I just realised that since we are going to set `withmpi` option in the code setup, and since it has higher priority than the plugin - there is no need to keep this option in the plugin.

Reply:
the code setup is made by the user of the plugin, we need a way to let the plugin developers to set the default mpi behavior of the code. But you are right, if the `withmpi` is a must set parameter for the code in code setup, it will always override the plugin set and leave it not being used. The is a problem.

Reply:
That is what I described in my last comment though: if there really is a conflict and a plugin can never be run with a specific value of `withmpi` and either the `Code` value or `metadata.option.withmpi` specifies the wrong value. The plugin can simply raise a validation error. The user will then simply have to change the input, or change `Code` that has the correct withmpi setting.